File: ds/lru.cc
#include <iostream>
#include <list>
#include <unordered_map>
#include <memory>
#include <utility>


// suppport set, get value with max capacity, least recently used items will be remove when capacity reached

class LRUCache {
    public:
        LRUCache(int x): capacity(x) {};

        void put(int key, int val);

        int get(int key);

    private:
        typedef std::list<int> LI;
        typedef std::pair<int, LI::iterator> PI;
        typedef std::unordered_map<int, PI> MPI;

        int capacity;
        MPI kv;
        LI ll;
};


void LRUCache::put(int key, int val) {
    // check if key exists
    // if exists, update value and order of linked list
    if (kv.find(key) != kv.end()) {
        auto it = kv[key].second;
        ll.erase(it);
    }

    // check capacity
    // if exceeds, remove least used item
    if (ll.size() == capacity) {
        int k = ll.back();
        ll.pop_back();
        kv.erase(k);
    }
    ll.push_front(key);
    kv[key] = {val, ll.begin()};
}

int LRUCache::get(int key) {
    if (kv.find(key) == kv.end()) return -1;
    auto p = kv[key];
    auto it = p.second;
    ll.erase(it);
    ll.push_front(key);
    kv[key] = {p.first, ll.begin()};
    return p.first;
}


int main() {
    int capacity = 10;
    auto lru = new LRUCache(10);

    for (int i = 0; i < 20; i++) {
        lru->put(i, i);
    }

    for (int i = 0; i < 20; i++) {
        std::cout << lru->get(i) << "\t";
    }
    delete lru;

    return 0;
}

File: golang/lru/lru.go
package main

import "container/list"

type Pair struct {
	first  int
	second *list.Element
}

type LRUCache struct {
	c int
	m map[int]Pair
	l *list.List
}

func NewLRUCache(c int) *LRUCache {
	return &LRUCache{
		c: c,
		m: make(map[int]Pair),
		l: list.New(),
	}
}

// Put will add/update value to cache in O(1)
func (c *LRUCache) Put(int key, int val) {
	if p, ok := m[key]; ok {
		c.l.Remove(p.second)
	}
	if c.c == c.l.Len() {
		c.l.Remove(c.l.Front())
	}
	c.l.PushBack(key)
	u := Pair{first: val, second: c.l.Back()}
	c.m[key] = u
}

func (c *LRUCache) Get(int key) int {
	if p, ok := m[key]; !ok {
		return -1
	} else {
		c.l.Remove(p.second)
		c.l.PushBack(key)
		u := Pair{first: p.first, second: c.l.Back()}
		c.m[key] = u
		return p.first
	}
}
